File: Modder/modder/src/main/java/modder/Adb.java
/* 
 * wrapper for running adb command
*/
package modder;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Adb {
    public enum Error {
        ok,
        no_connection,
    }

    public class Output {

        public List<String> strings = new ArrayList<String>();
        public Error error = Error.ok;

    }

    public Output Run(List<String> command) {
        Output out = new Output();

        // test if adb shell can be connected by echoing in
        // the shell and check the output
        out.strings = Util.RunCommand("adb", Arrays.asList("shell", "echo", "test"));
        if (out.strings.size() >= 1) {
            if (!out.strings.get(0).equals("test"))
                out.error = Error.no_connection;

        }

        else {
            out.error = Error.no_connection;
        }

        // shell is alive, execute shell command
        if (out.error == Error.ok) {
            out.strings = Util.RunCommand("adb", command);
        }
        return out;

    }

    /*
     * Run a command inside adb shell
     * before running [command], it will check if the shell can
     * be connected by trying to run "echo test" and check its output
     * 
     * 
     */
    public Output RunShell(List<String> shellArg) {
        List<String> commands = new ArrayList<String>();
        //
        commands.add("shell");
        for (String s : shellArg)
            commands.add(s);
        return Run(commands);

    }

    public Output ListApk() {

        Adb.Output out = RunShell(Arrays.asList("pm", "list", "packages"));
        if (out.error != Adb.Error.ok) {
            return out;
        }
        List<String> apks = new ArrayList<String>();
        // output will look like
        // package:com.android.offfice
        // package:com.vivo.appstore
        // "package:" should be trimmed for better view
        for (int i = 0; i < out.strings.size(); i++) {
            // use the caret symbol '^'
            // to match the beggining of the pattern
            String new_str = out.strings.get(i).replaceFirst("^package:", "");
            apks.add(new_str);
        }
        out.strings = apks;
        return out;
    }

    public Output GetApkPathAtDevice(String package_name) {

        Adb.Output out = RunShell(Arrays.asList("pm", "path", package_name));
        if (out.error != Adb.Error.ok) {
            return out;
        }
        List<String> apks_paths = new ArrayList<String>();
        // output will look like
        // package:/system/app/name
        // package:/system/test/clashroyale
        // "package:" should be trimmed
        for (int i = 0; i < out.strings.size(); i++) {
            // use the caret symbol '^'
            // to match the beggining of the pattern
            String new_str = out.strings.get(i).replaceFirst("^package:", "");
            apks_paths.add(new_str);
        }
        out.strings = apks_paths;
        return out;
    }

    public Output DownloadApk(String path_to_apk, String downloadDir) {

        Output out = Run(Arrays.asList("pull", path_to_apk, downloadDir));
        if (out.error != Adb.Error.ok) {
            return out;
        }
        return out;
    }

    public Output InstallApk(String apkDirStr) throws IOException {

        File apkDir = new File(apkDirStr);
        Assert.AssertExistAndIsDirectory(apkDir);
        List<String> command = new ArrayList<String>();
        // use install-multiple to install one or more apk
        // https://android.stackexchange.com/questions/221204/how-to-install-xapk-apks-or-multiple-apks-via-adb
        command.add("install-multiple");
        for (File f : apkDir.listFiles()) {
            // only collect file that has extension of *.apk
            if (f.getAbsolutePath().endsWith(".apk"))
                command.add(f.getAbsolutePath());
        }
        Output out = Run(command);
        return out;

    }

}

File: Modder/modder/src/main/java/modder/Adb.kt
/* 
 * wrapper for running adb command
*/
package modder

import java.io.*
import java.util.*

class Adb {
    enum class Error {
        ok,
        no_connection
    }

    inner class Output {
        var strings: List<String> = ArrayList()
        var error = Error.ok
    }

    fun Run(command: List<String>): Output {
        val out: Output = Output()

        // test if adb shell can be connected by echoing in
        // the shell and check the output
        out.strings = Util.RunCommand("adb", mutableListOf<String>("shell", "echo", "test"))
        if (out.strings.size >= 1) {
            if (out.strings.get(0) != "test") out.error = Error.no_connection
        } else {
            out.error = Error.no_connection
        }

        // shell is alive, execute shell command
        if (out.error == Error.ok) {
            out.strings = Util.RunCommand("adb", command)
        }
        return out
    }

    /*
     * Run a command inside adb shell
     * before running [command], it will check if the shell can
     * be connected by trying to run "echo test" and check its output
     * 
     * 
     */
    fun RunShell(shellArg: List<String>): Output {
        val commands: MutableList<String> = ArrayList()
        //
        commands.add("shell")
        for (s in shellArg) commands.add(s)
        return Run(commands)
    }

    fun ListApk(): Output {
        val out = RunShell(mutableListOf<String>("pm", "list", "packages"))
        if (out.error != Error.ok) {
            return out
        }
        val apks: MutableList<String> = ArrayList()
        // output will look like
        // package:com.android.offfice
        // package:com.vivo.appstore
        // "package:" should be trimmed for better view
        for (i in out.strings!!.indices) {
            // use the caret symbol '^'
            // to match the beggining of the pattern
            val new_str = out.strings!![i]!!.replaceFirst("^package:".toRegex(), "")
            apks.add(new_str)
        }
        out.strings = apks
        return out
    }

    fun GetApkPathAtDevice(package_name: String): Output {
        val out = RunShell(Arrays.asList("pm", "path", package_name))
        if (out.error != Error.ok) {
            return out
        }
        val apks_paths: MutableList<String> = ArrayList()
        // output will look like
        // package:/system/app/name
        // package:/system/test/clashroyale
        // "package:" should be trimmed
        for (i in out.strings!!.indices) {
            // use the caret symbol '^'
            // to match the beggining of the pattern
            val new_str = out.strings!![i]!!.replaceFirst("^package:".toRegex(), "")
            apks_paths.add(new_str)
        }
        out.strings = apks_paths
        return out
    }

    fun DownloadApk(path_to_apk: String, downloadDir: String): Output {
        val out = Run(Arrays.asList("pull", path_to_apk, downloadDir))
        return if (out.error != Error.ok) {
            out
        } else out
    }

    @Throws(IOException::class)
    fun InstallApk(apkDirStr: String): Output {
        val apkDir = File(apkDirStr)
        Assert.AssertExistAndIsDirectory(apkDir)
        val command: MutableList<String> = ArrayList()
        // use install-multiple to install one or more apk
        // https://android.stackexchange.com/questions/221204/how-to-install-xapk-apks-or-multiple-apks-via-adb
        command.add("install-multiple")
        for (f in apkDir.listFiles()) {
            // only collect file that has extension of *.apk
            if (f.absolutePath.endsWith(".apk")) command.add(f.absolutePath)
        }
        return Run(command)
    }
}
